### ISSUE TITLE
Linux: Tasks not using $PATH from ~/.bashrc

### ISSUE THREAD

**Issue by TheColorRed** [Microsoft/vscode#29412]
- VSCode Version: Code 1.13.1 (379d2efb5539b09112c793d3d9a413017d736f89, 2017-06-14T18:18:03.540Z)
- OS Version: Linux x64 4.10.0-24-generic

---

I have a custom location for my global npm modules `~/.npm-global`, and I have the bin location set in my `.bashrc` like so ([from npm doc](https://docs.npmjs.com/getting-started/fixing-npm-permissions#option-2-change-npms-default-directory-to-another-directory)):

```
export PATH="/home/ryan/.npm-global/bin:$PATH"
```

When executing `which node-sass` in the vscode terminal I get `/home/ryan/.npm-global/bin/node-sass` which is correct, but when I attempt to run from my `.vscode/tasks.json` file I get an error:

>/bin/bash: node-sass: command not found

The is the task file I am using:

```json
{
  "version": "2.0.0",
  "isShellCommand": true,
  "showOutput": "always",
  "isBackground": true,
  "echoCommand": true,
  "tasks": [
    {
      "taskName": "Build Client Sass",
      "command": "node-sass",
      "args": [
        "-w",
        "./src/client/sass/main.scss",
        "./dist/client/assets/css/media-server.min.css",
        "--output-style",
        "compressed"
      ]
    }
  ]
}
```

Once I change `"command": "node-sass"` to `"command": "/home/ryan/.npm-global/bin/node-sass"` then it works.

Basically, I don't think that the tasks are loading the `$PATH` information correctly...

**Comment by dbaeumer** [comment]
The problem here is the bash behavior when it comes to run bash -c. Most .bashrc files have a section like this at the top which 

```
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
```

which exists the .bashrc profile if not run in interactive mode. You can avoid this by running bash -ic when executing a command but we stayed away from making this the default since not everybody wants to resource the .bashrc. 

What is strange here is the fact that the PATH should be correct to begin with when running VS Code and then the correct PATH should travel to the bash -c command.

@TheColorRed how do you start code. Does the same happen when you start code from a terminal ?

**Comment by TheColorRed** [comment]
@dbaeumer I have it saved in my favorites, and with the extension `Dash to dock` it is pinned to the dock. I then just click on it from there.

When I open it from the terminal using `code .` and run the task without the full path to the bin file: `node-sass -w -o ./assets/css -- ./sass` it does in fact work. When opening from my favorites, it does not work unless it is the full path.

**Comment by dbaeumer** [comment]
@TheColorRed thanks. This means that the environment is not fully sourced when starting VS Code from the dock. One more question: which shell is your default shell? bash?

@joaomoreno assigning this to you as well since it looks like a bug with sourcing the initial environment in VS Code when started from the dock.

**Comment by joaomoreno** [comment]
If you open Code not from the terminal, and open the Developer Tools, is there any message in the Console?

**Comment by TheColorRed** [comment]
@dbaeumer when running `echo $0` the output is `bash`.

@joaomoreno There are messages, however, they don't seem related to the issue as they are the same messages that appear when opening from outside of the terminal

**Comment by joaomoreno** [comment]
In the Devtools console, if you type process.env, does it look incomplete to you?

**Comment by TheColorRed** [comment]
I can tell you that the only difference is that when starting from the terminal the `process.env.PATH` contains `/home/ryan/.npm-global/bin` whereas starting it from my dock `process.env.PATH` does not contain that value

**Comment by joaomoreno** [comment]
Is the code that adds that path to `PATH` in `.bashrc` surrounded by some `if` condition?

**Comment by TheColorRed** [comment]
nope, it is the last line of the file

**Comment by joaomoreno** [comment]
Can you show me the file?

**Comment by TheColorRed** [comment]
However, the `~/.profile` wraps it within an if:

```
if [ -n "$BASH_VERSION" ]; then
    # include .bashrc if it exists
    if [ -f "$HOME/.bashrc" ]; then
        . "$HOME/.bashrc"
    fi
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/bin" ] ; then
    PATH="$HOME/bin:$PATH"
fi
```

Here is the **.bashrc**
```
# ~/.bashrc: executed by bash(1) for non-login shells.
# see /usr/share/doc/bash/examples/startup-files (in the package bash-doc)
# for examples

# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac

# don't put duplicate lines or lines starting with space in the history.
# See bash(1) for more options
HISTCONTROL=ignoreboth

# append to the history file, don't overwrite it
shopt -s histappend

# for setting history length see HISTSIZE and HISTFILESIZE in bash(1)
HISTSIZE=1000
HISTFILESIZE=2000

# check the window size after each command and, if necessary,
# update the values of LINES and COLUMNS.
shopt -s checkwinsize

# If set, the pattern "**" used in a pathname expansion context will
# match all files and zero or more directories and subdirectories.
#shopt -s globstar

# make less more friendly for non-text input files, see lesspipe(1)
[ -x /usr/bin/lesspipe ] && eval "$(SHELL=/bin/sh lesspipe)"

# set variable identifying the chroot you work in (used in the prompt below)
if [ -z "${debian_chroot:-}" ] && [ -r /etc/debian_chroot ]; then
    debian_chroot=$(cat /etc/debian_chroot)
fi

# set a fancy prompt (non-color, unless we know we "want" color)
case "$TERM" in
    xterm-color) color_prompt=yes;;
esac

# uncomment for a colored prompt, if the terminal has the capability; turned
# off by default to not distract the user: the focus in a terminal window
# should be on the output of commands, not on the prompt
force_color_prompt=yes

if [ -n "$force_color_prompt" ]; then
    if [ -x /usr/bin/tput ] && tput setaf 1 >&/dev/null; then
	# We have color support; assume it's compliant with Ecma-48
	# (ISO/IEC-6429). (Lack of such support is extremely rare, and such
	# a case would tend to support setf rather than setaf.)
	color_prompt=yes
    else
	color_prompt=
    fi
fi

if [ "$color_prompt" = yes ]; then
    PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
else
    PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w\$ '
fi
unset color_prompt force_color_prompt

# If this is an xterm set the title to user@host:dir
case "$TERM" in
xterm*|rxvt*)
    PS1="\[\e]0;${debian_chroot:+($debian_chroot)}\u@\h: \w\a\]$PS1"
    ;;
*)
    ;;
esac

# enable color support of ls and also add handy aliases
if [ -x /usr/bin/dircolors ]; then
    test -r ~/.dircolors && eval "$(dircolors -b ~/.dircolors)" || eval "$(dircolors -b)"
    alias ls='ls --color=auto'
    #alias dir='dir --color=auto'
    #alias vdir='vdir --color=auto'

    alias grep='grep --color=auto'
    alias fgrep='fgrep --color=auto'
    alias egrep='egrep --color=auto'
fi

# some more ls aliases
alias ll='ls -alF'
alias la='ls -A'
alias l='ls -CF'

# Add an "alert" alias for long running commands.  Use like so:
#   sleep 10; alert
alias alert='notify-send --urgency=low -i "$([ $? = 0 ] && echo terminal || echo error)" "$(history|tail -n1|sed -e '\''s/^\s*[0-9]\+\s*//;s/[;&|]\s*alert$//'\'')"'

# Alias definitions.
# You may want to put all your additions into a separate file like
# ~/.bash_aliases, instead of adding them here directly.
# See /usr/share/doc/bash-doc/examples in the bash-doc package.

if [ -f ~/.bash_aliases ]; then
    . ~/.bash_aliases
fi

# enable programmable completion features (you don't need to enable
# this, if it's already enabled in /etc/bash.bashrc and /etc/profile
# sources /etc/bash.bashrc).
if ! shopt -oq posix; then
  if [ -f /usr/share/bash-completion/bash_completion ]; then
    . /usr/share/bash-completion/bash_completion
  elif [ -f /etc/bash_completion ]; then
    . /etc/bash_completion
  fi
fi

[ -s "/home/ryan/.dnx/dnvm/dnvm.sh" ] && . "/home/ryan/.dnx/dnvm/dnvm.sh" # Load dnvm

parse_git_branch() {
     git branch 2> /dev/null | sed -e '/^[^*]/d' -e 's/* \(.*\)/(\1)/'
}
if [ "$color_prompt" = yes ]; then
      PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[01;31m\]$(parse_git_branch)\[\033[00m\]\$ '
  else
       PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w$(parse_git_branch)\$ '
   fi
   unset color_prompt force_color_prompt

###-tns-completion-start-###
if [ -f /home/ryan/.tnsrc ]; then 
    source /home/ryan/.tnsrc 
fi
###-tns-completion-end-###
ANDROID_HOME=/home/ryan/Android/Sdk/
JAVA_HOME=/usr/lib/jvm/java-8-oracle
wine32="WINEPREFIX='/home/ryan/.wine32' wine"
export PATH="/home/ryan/.npm-global/bin:$PATH"
```

**Comment by joaomoreno** [comment]
The problem is right at the beginning of that file:

```
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
```

#27185 #21655

Not really sure how to fix that.

**Comment by dbaeumer** [comment]
@joaomoreno for bash you can do the following when executing the command to source the env into VS code `bash -ic`. This executes the command and pretends to run in interactive mode and even exits afterwards. Not sure if this works with other shells.

**Comment by joaomoreno** [comment]
@dbaeumer I've been there many times and broken this mechanism many times due to it. I won't try again.

**Comment by azurvii** [comment]
Not sure how I can reopen this issue, but it looks like it does not "need more information", but needs a solution.
If changing how Code is launched is complicated, how about changing how Code launches tools? e.g. adding an option in settings.json to allow search additional paths?
